subservers/manager.go

```diff
@@ -0,0 +1,265 @@
+	})
+}
+
+// StartIntegratedServers starts all the manager's sub-servers that should be
+// started in integrated mode.
+func (s *Manager) StartIntegratedServers(lndClient lnrpc.LightningClient,
+	lndGrpc *lndclient.GrpcLndServices, withMacaroonService bool) {
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, ss := range s.servers {
+		if ss.Remote() {
+			continue
+		}
+
+		err := ss.startIntegrated(
+			lndClient, lndGrpc, withMacaroonService,
+		)
+		if err != nil {
+			continue
+		}
+	}
+}
+
+// ConnectRemoteSubServers creates connections to all the manager's sub-servers
+// that are running remotely.
+func (s *Manager) ConnectRemoteSubServers() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, ss := range s.servers {
+		if !ss.Remote() {
+			continue
+		}
+
+		err := ss.connectRemote()
+		if err != nil {
+			log.Errorf("Failed to connect to remote %s: %v",
+				ss.Name(), err)
+
+			continue
+		}
+	}
+}
+
+// RegisterRPCServices registers all the manager's sub-servers with the given
+// grpc registrar.
+func (s *Manager) RegisterRPCServices(server grpc.ServiceRegistrar) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, ss := range s.servers {
+		// In remote mode the "director" of the RPC proxy will act as
+		// a catch-all for any gRPC request that isn't known because we
+		// didn't register any server for it. The director will then
+		// forward the request to the remote service.
+		if ss.Remote() {
+			continue
+		}
+
+		ss.RegisterGrpcService(server)
+	}
+}
+
+// GetRemoteConn checks if any of the manager's sub-servers owns the given uri
+// and if so, the remote connection to that sub-server is returned. The bool
+// return value indicates if the uri is managed by one of the sub-servers
+// running in remote mode.
+func (s *Manager) GetRemoteConn(uri string) (bool, *grpc.ClientConn) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, ss := range s.servers {
+		// TODO(positiveblue): check subserver permissions.
+
+		if !ss.Remote() {
+			return false, nil
+		}
+
+		return true, ss.remoteConn
+	}
+
+	return false, nil
+}
+
+// ValidateMacaroon checks if any of the manager's sub-servers owns the given
+// uri and if so, if it is running in remote mode, then true is returned since
+// the macaroon will be validated by the remote subserver itself when the
+// request arrives. Otherwise, the integrated sub-server's validator validates
+// the macaroon.
+func (s *Manager) ValidateMacaroon(ctx context.Context,
+	requiredPermissions []bakery.Op, uri string) (bool, error) {
+
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, ss := range s.servers {
+		// TODO(positiveblue): check subserver permissions.
+
+		if ss.Remote() {
+			return true, nil
+		}
+
+		if !ss.started() {
+			return true, fmt.Errorf("%s is not yet ready for "+
+				"requests, lnd possibly still starting or "+
+				"syncing", ss.Name())
+		}
+
+		err := ss.ValidateMacaroon(ctx, requiredPermissions, uri)
+		if err != nil {
+			return true, fmt.Errorf("invalid macaroon: %v", err)
```

Comment:
It feels like this should return `false`, just to be on the safe side. And then I guess we need a `return true` at the end, since all the validation succeeded if there was no error?

Reply:
Sorry for the misleading comment here... This is now resolved correctly though.